### ISSUE TITLE
RPC format relies on eval for large payloads and long strings

### ISSUE THREAD

**Issue by niloc132** [gwtproject/gwt#9578]
**GWT version:2.8.2** 
**Browser (with version):any** 
**Operating System:any** 

---

##### Description
From https://github.com/gwtproject/gwt/issues/8197#issuecomment-349342472, it turns out that RPC's server serialization stream writer still assumes that the browser needs concatenated strings and arrays. 

However, the strings were concatenated to support hosted mode (which apparently at the time couldn't support larger than 64kb strings), but that appears to have been fixed, and arrays were split to support a bug in IE6/7 where long arrays couldn't be eval'd, but again, we don't support IE6/7 any more.

So, we don't really need it. Additionally, this is an issue since it violates CSP, as eval isn't particularly safe, and some sites would like to forbid its use to further protect their data and their users.

I'll try removing these hacks and confirm that eval is no longer called (except of course in browsers that do not support `JSON.parse`), and confirm that running dev mode can handle giant strings correctly as well. This should remove the ability of the server to produce streams older than version 8 (since `ServerSerializationStreamWriter.writeHeader` is the only code that should be writing out a version for the client). Will leave in support on the client for reading older streams, as we've done in the past.

**Comment by heartdisease** [comment]
Any update on this? We are using GWT at work and would love to finally get rid of the `unsafe-eval` rule.

**Comment by niloc132** [comment]
I do have this patch somewhere, but can't recall where it stalled, probably in making sure that other browsers tested well.

Instead, most of my time in this area has been spent on a feature-complete replacement for RPC, based on APT instead of Generators. This is at https://github.com/vertispan/gwt-rpc/, and while it isn't ever going to get to 100%, it is probably at about 70% now, and should be able to reach 80-90% (minus just a few features like reflection/violator pattern to modify private or final fields directly, so bean-like getters and setters or custom field serializers will be required).

I'll see if I can find the patch, but this work was not needed for the new RPC at all, which doesn't mess with json strings at all at this time.

**Comment by heartdisease** [comment]
So that'd mean that without this patch we'd be kind of stuck with `unsafe-eval` unless we upgrade to GWT 3?

**Comment by niloc132** [comment]
No - while the linked project is intended to be compatible with GWT 3, it is already compatible with GWT 2 (and that isn't expected to change).

**Comment by Zim123** [comment]
Hello.

I'm not sure if this is the correct place for my question, but it was the only place more relative to what I'm struggling with. And to not open a new issue with relatively the same problem.

So, my problem is related to CSP, specifically 'unsafe-inline' rule, because now, in my situation compiled application violates it. Whether I'm using 'xsiframe' or 'direct_install' linker, the situation is related to the fact I'm using deferred split modules, which are injecting via inline js script to iframe, and in this way violate the 'unsave-inline' rule. I could use sso linker, but it's not a case, because it does not support several modules.

Can anyone suggest maybe some linker or approach, to fix this issue, please?

Regards.

**Comment by tbroyer** [comment]
@Zim123 Let's discuss on #8197 or, better, https://groups.google.com/forum/#!forum/google-web-toolkit

**Comment by bcampolo** [comment]
We are hitting this issue (9578) as well.  We recently had a security scan find the use of unsafe-eval in our GWT-based application.  Upon removing unsafe-eval and setting the gwt.rpc.version to 8, everything worked except for requests with large responses > 32k, which executed an eval instead of safeEval.  It appears that ServerSerializationStreamWriter.LengthConstrainedArray is used to conditionally change the response when the size is greater than 32k which resets the rpc version back to 7 and uses a special "PRELUDE" concatenator and flags the response as 'javascript'.  There are comments in this thread and in the code which indicates this is only needed for IE 6/7 both of which are no longer supported.  We quickly built GWT locally to test with larger size constraint and this fixed the issue we were seeing.  Do you know if a fix for this issue would be included in the next release?

Section of code that I referenced regarding the length check:
https://github.com/gwtproject/gwt/blob/main/user/src/com/google/gwt/user/server/rpc/impl/ServerSerializationStreamWriter.java#L71

**Comment by niloc132** [comment]
Yes, I think we could probably remove that workaround entirely - are you interested in putting together a patch for it?

**Comment by bcampolo** [comment]
I appreciate your willingness to accept a patch.  Unfortunately my company is not able to prioritize the work to create a patch at this time.  I will monitor this issue and let you know if that changes.

**Comment by hijackit** [comment]
how do you set gwt.rpc.version to version 8?

**Comment by niloc132** [comment]
Sorry, this isn't fixed, but the workaround is improved.

**Comment by Lonzak** [comment]
We are also running into this issue - has this been patched by someone? bcampolo was describing it and there were some commits by codemasterover9000. 

> We should raise these limits to their max and update the default version to 8 for the next release

Did that happen and which release is it?

**Comment by niloc132** [comment]
@Lonzak the patches merged for 2.12 already should let you configure these size limits - set them to a huge value, and you'll never see this issue again. 

We will raise the defaults and remove the ancient IE specific implementations - that work is pending still in https://github.com/gwtproject/gwt/pull/9876, but someone else could make a new patch to supersede it.

**Comment by Lonzak** [comment]
> [@Lonzak](https://github.com/Lonzak) the patches merged for 2.12 already should let you configure these size limits - set them to a huge value, and you'll never see this issue again.
> We will raise the defaults and remove the ancient IE specific implementations - that work is pending still in [#9876](https://github.com/gwtproject/gwt/pull/9876), but someone else could make a new patch to supersede it.

Interesting browsed the other ticket - is it really time to remove IE support / code. What about rhino and why is an issue? Is it the Mozilla Rhino javascript engine (used in java 6 and 7) or what is the problem?

**Comment by niloc132** [comment]
Rhino is a Java implementation of JS, and GWT uses it for htmlunit testing, and has a fork of it for jsni parsing and minifying/compiling linker output. I believe the particular part that is failing is specific to the fork that I mentioned, and how that is additionally used in legacy dev mode to avoid an expensive call into JS. It probably could be replaced with a JSON parser instead of the minimum rpc version is set to 8 so that JSON is always used.

Removing IE is generally a good idea - much of the IE support has been removed where it could be cleanly untangled from what other browsers do. In this case as you can see from the linked patch, this isn't just a matter of removing some rebind rules and IE-specific implementations, but dealing with the other assumptions and checks made so that we don't need all the extra code to handle those cases in either the client or the server. This is a little more complicated than just deleting the "is this IE" checks, so hasn't been done yet.

**Comment by Lonzak** [comment]
Thank you for the explanation and your insight. I can imagine that something like this is not easily untangled in all places ... 
One question regarding rhino - what do you think - wouldn't it be possible to update the GWT rhino fork with a newer version where the issue (and others) were fixed or is it better to completely get rid of it by replacing it with something else? Or maybe two step approach?

**Comment by niloc132** [comment]
The fork isn't to handle the max string size limit, it instead is to add support when parsing JSNI for methods to java references - see my note in the PR for more info.

Killing legacy dev mode or using a proper java JSON library would be better answers for that failing test.
